src/Cli/dotnet/ProjectLocator.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.Build.Execution;
+using Microsoft.DotNet.Cli.Utils;
+
+namespace Microsoft.DotNet.Cli
+{
+    abstract class ProjectLocator
+    {
+        /// <returns>A project instance that will be targeted to publish/pack, etc. null if one does not exist.</returns>
+        public abstract ProjectInstance GetTargetedProject(IEnumerable<string> slnOrProjectArgs, string slnProjectPropertytoCheck = "");
+
+        /// <returns>The top-level project (first if multiple exist) in a SLN. Returns null if no top level project. Throws exception if two top level projects disagree
+        /// in the configuration property to check.</returns>
+        public abstract ProjectInstance GetSlnProject(string potentialSlnPath, string slnProjectConfigPropertytoCheck = "");
+
+        /// <returns>Creates a ProjectInstance if the project is valid, elsewise, fails..</returns>
+        protected static ProjectInstance TryGetProjectInstance(string projectPath)
+        {
+            try
+            {
+                return new ProjectInstance(projectPath);
```

Comment:
We should probably pass through any global properties that are going to be passed to MSBuild.  That will help us correctly handle things like `--property:Configuration=Release`, but also something like `--property:PublishRelease=true` or other weirder things.

This is another scenario to add tests for, too.

Reply:
Didn't consider what to do if the project defines `PublishRelease` `true` but its given `PublishRelease` `false` through `property:`. I would rather it respect the CLI over the project personally. But maybe we should just error.

Reply:
Looks like for all other properties CLI takes precedence. So gonna do that.